### PR TITLE
Use automatic detection of environment for hardware concurrency

### LINKED ISSUE #1152: libSquoosh declares a global `navigator` variable
**Before you start**
Please take a look at the [FAQ](https://github.com/GoogleChromeLabs/squoosh/wiki/FAQ) as well as the already opened issues! If nothing fits your problem, go ahead and fill out the following template:

**Describe the bug**

`import "@squoosh/lib"` add a global `navigator` variable in `codecs.ts`. The issue is that some libraries assume the code is running is a web browser environment if there's a `navigator` object available on the global object.

**To Reproduce**

```sh
mkdir repro
cd repro
npm i sass @squoosh/lib
echo 'import "@squoosh/lib"; import "sass";' > test.mjs
node ./test.mjs
```

**Expected behavior**

`test.mjs` should exit with code 0.

**Actual behavior**

```
…/repro/node_modules/sass/sass.dart.js:92529
    if (ua.indexOf("DumpRenderTree") >= 0) return hooks;
           ^

TypeError: Cannot read properties of undefined (reading 'indexOf')
    at …/repro/node_modules/sass/sass.dart.js:92529:12
    at Object.applyHooksTransformer (…/repro/node_modules/sass/sass.dart.js:7236:14)
    at Object.initHooks (…/repro/node_modules/sass/sass.dart.js:7216:251)
    at Object.initNativeDispatchContinue (…/repro/node_modules/sass/sass.dart.js:7182:9)
    at Object.initNativeDispatch (…/repro/node_modules/sass/sass.dart.js:7176:9)
    at Object.getNativeInterceptor (…/repro/node_modules/sass/sass.dart.js:9070:13)
    at Object.getInterceptor$x (…/repro/node_modules/sass/sass.dart.js:9326:16)
    at Object.set$render$x (…/repro/node_modules/sass/sass.dart.js:9363:16)
    at main0 (…/repro/node_modules/sass/sass.dart.js:17757:9)
    at …/repro/node_modules/sass/sass.dart.js:93924:7
```

**Version:**
 - OS w/ version: [e.g. iOS 12] N/A
 - Browser w/ version [e.g. Chrome 70] N/A
 - Node version: [e.g. 10.11.0] 16.9.1
 - npm version: [e.g. 6.4.1] 7.21.1

**Additional context, screenshots, screencasts**

It seems that Emscripten allows us to use anything for the pool size, so I suggest using something like `globalThis[Symbol.for('squoosh.navigator')].hardwareConcurrency` instead. I could also open a bug on the Sass issue tracker, but I don't think Sass is the only library that check for `globalThis.navigator` to make assumptions, and it'd be better to avoid it IMHO.

### PR DESCRIPTION
**Pull request by TimvdLippe**
We can pass in any string into `PTHREAD_POOL_SIZE`, which means we can
inject a bit of code to implement the switch-case for both Node and web.
On the web, it will use `navigator.hardwareConcurrency` as before, while
on Node it will use `cpus().length`. This allows us to remove the
assignment to `navigator` on the global scope in `codecs.ts`, which
avoids throwing off tools like `sass` which assume a web environment
when that global is set.

Fixes #1152
Builds upon #1174

### CODE REVIEW

File: codecs/cpp.Dockerfile [hunk 1]
```diff
@@ -4,7 +4,7 @@ ENV CFLAGS "-O3 -flto"
 ENV CXXFLAGS "${CFLAGS} -std=c++17"
 ENV LDFLAGS "${CFLAGS} \
 -s FILESYSTEM=0 \
--s PTHREAD_POOL_SIZE=navigator.hardwareConcurrency \
+-s PTHREAD_POOL_SIZE=\"ENVIRONMENT_IS_NODE?require('os').cpus().length:navigator.hardwareConcurrency\" \
```
Comment (TimvdLippe) [inline codecs/cpp.Dockerfile:7]: Note that I tried to keep this with spaces, but unfortunately appeasing both Bash and Docker didn't seem to work out. Instead, we can remove the spaces (which are only visual anyways) and avoid these spacing issues in the environment argument.

Comment (RReverser) [reply codecs/cpp.Dockerfile:7]: Huh weird - I'd expect it to handle spaces with escaped quotes like any other - but okay.

File: codecs/build-all.sh [hunk 2]
```diff
@@ -0,0 +1,8 @@
+CURRENT_DIRECTORY=$(dirname "$(readlink -f "$0")")
+SUB_DIRECTORIES=$(find "$CURRENT_DIRECTORY" -mindepth 1 -maxdepth 1 -type d)
```
Comment (RReverser) [inline codecs/build-all.sh:2]: Nit: might be simpler to use `ls -d $CURRENT_DIRECTORY/*/`.

Comment (TimvdLippe) [reply codecs/build-all.sh:2]: Changed in #1174

**Comment by RReverser** [comment]
Hm not sure why we didn't merge this. @surma Thoughts?

**Comment by donmccurdy** [comment]
This would help me as well – I currently can't use `@squoosh/lib` alongside another package that does something like:

```js
if (typeof navigator === 'undefined') return;
// proceeds to assume Web environment ...
```

**Comment by donmccurdy** [comment]
@RReverser @jakearchibald is there anything I could do to help get this PR merged? E.g. rebasing it or testing some specific use cases? I'd like to use `@squoosh/lib` in a project but the modification of the global namespace (below) is breaking other dependencies that use `navigator` to detect Web vs. Browser contexts.

```js
// breaks other project dependencies
(globalThis as any).navigator = {
  hardwareConcurrency: cpus().length,
};
```

**Comment by burner03** [comment]
I just run 

`navigator = '';` 

at the top of my automation. Not a great solution. Would love to see this merged.

**Comment by donmccurdy** [comment]
For anyone else blocked – I had another package breaking on the navigator shim. My workaround was similar to what @burner03 describes, but I had to replace async 'import' with synchronous 'require' to ensure the fix was in place at the right time. This got a little trickier with TypeScript, so here's the full workaround:

before:

```js
import other from 'other-package';
```

after:

```typescript
import type Other from 'other-package';

const other = createOther();

/** Creates and returns a 'other' instance. */
function createOther(): typeof Other {
	// Workaround for https://github.com/GoogleChromeLabs/squoosh/pull/1176.
	const navigator = global.navigator;
	delete (global as Record<string, unknown>).navigator;
	try {
		return require('other-package');
	} finally {
		global.navigator = navigator;
	}
}
```

**Pull request closed without merging**